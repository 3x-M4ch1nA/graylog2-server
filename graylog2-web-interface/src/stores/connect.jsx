import React from 'react';
import Reflux from 'reflux';
import createClass from 'create-react-class';

import Spinner from 'components/common/Spinner';

/**
 * Generating a higher order component wrapping an ES6 React component class, connecting it to the supplied stores.
 * The generated wrapper class passes the state it receives from the stores to the component it wraps as props.
 *
 * The `stores` parameter should consist of an object mapping desired props keys to stores. i.e.:
 * {
 *  samples: SamplesStore
 * }
 *
 * will connect to the provided `SamplesStore` and pass the state it receives from it to the supplied component as a
 * `samples` prop.
 *
 * Upon mount of the generated component all functions included in `initialActions` will be executed. This is usable to
 * execute actions on one or more stores, which populate the store(s).
 *
 * When either `initialActions` is non-empty or `waitForStores` is set to `true`, the generated component will return a
 * spinner instead of the supplied component if at least one of the mapped props keys is `undefined` (meaning that at
 * least one state of the stores is `undefined`, usually meaning that it is not available yet).
 *
 * @param {Object} Component - The component which should be connected to the stores.
 * @param {Object} stores - A mapping of desired props keys to stores.
 * @param {function[]} initialActions - An array of functions being called when the generated component is mounted.
 * @param {boolean} [waitForStores=false] - Defines if the generated component waits for all specified keys in its state
 *                                        before rendering `Component`.
 * @returns {Object} - A React component wrapping the passed component.
 *
 * @example
 *
 * // Connecting the `SamplesComponent` class to the `SamplesStore`, hooking up its state to the `samples` prop and
 * // waiting for the store to settle down.
 *
 * connect(SamplesComponent, { samples: SamplesStore }, [SamplesActions.list])
 *
 */
export default function connect(Component, stores = {}, initialActions = [], waitForStores = false) {
  const storeMixins = [];
  Object.keys(stores).forEach((key) => {
    storeMixins.push(Reflux.connect(stores[key], key));
  });

  const StoreConnectionWrapper = createClass({
    mixins: storeMixins,
    componentDidMount() {
      initialActions.forEach(action => action());
    },
    render() {
      const props = {};
      Object.keys(stores).forEach((key) => {
        props[key] = this.state[key];
      });

      if (waitForStores || initialActions.length > 0) {
        if (Object.keys(stores).map((key) => { return this.state[key] !== undefined; }).find(elem => elem === false) !== undefined) {
          return <Spinner />;
        }
      }
      return <Component {...props} />;
    },
  });

  return StoreConnectionWrapper;
}
