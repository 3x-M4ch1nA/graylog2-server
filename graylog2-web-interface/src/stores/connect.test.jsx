/* eslint-disable react/no-multi-comp */
import React from 'react';
import renderer from 'react-test-renderer';
import Reflux from 'reflux';
import PropTypes from 'prop-types';

import connect from './connect';

class SimpleComponentWithoutStores extends React.Component {
  render() {
    return <span>Hello World!</span>;
  }
}

const SimpleStore = Reflux.createStore({
  getInitialState() {
    return this.state;
  },
  setValue(value) {
    this.state = { value: value };
    this.trigger(this.state);
  },
  noop() {},
  reset() {
    this.state = undefined;
    this.trigger(this.state);
  },
});

class SimpleComponentWithDummyStore extends React.Component {
  static propTypes = {
    simpleStore: PropTypes.shape({ foo: PropTypes.number }),
  };
  static defaultProps = {
    simpleStore: undefined,
  };

  render() {
    const { simpleStore } = this.props;
    if (simpleStore && simpleStore.value) {
      return <span>Value is: {simpleStore.value}</span>;
    }
    return <span>No value.</span>;
  }
}

describe('connect()', () => {
  it('does not do anything if no stores are provided', () => {
    const Component = connect(SimpleComponentWithoutStores, {});
    const wrapper = renderer.create(<Component />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('connects component to store without state', () => {
    const Component = connect(SimpleComponentWithDummyStore, { simpleStore: SimpleStore });
    const wrapper = renderer.create(<Component />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('connects component to store with state', () => {
    SimpleStore.setValue(42);
    const Component = connect(SimpleComponentWithDummyStore, { simpleStore: SimpleStore });
    const wrapper = renderer.create(<Component />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('renders spinner when store is not ready', () => {
    SimpleStore.reset();
    const Component = connect(SimpleComponentWithDummyStore, { simpleStore: SimpleStore }, [], true);
    const wrapper = renderer.create(<Component />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('executes action on store upon mount', () => {
    SimpleStore.reset();
    const Component = connect(SimpleComponentWithDummyStore, { simpleStore: SimpleStore }, [() => SimpleStore.setValue(23)]);
    const wrapper = renderer.create(<Component />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('waits on store per default before rendering component when more than zero actions are passed', () => {
    SimpleStore.reset();
    const Component = connect(SimpleComponentWithDummyStore, { simpleStore: SimpleStore }, [SimpleStore.noop]);
    const wrapper = renderer.create(<Component />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
