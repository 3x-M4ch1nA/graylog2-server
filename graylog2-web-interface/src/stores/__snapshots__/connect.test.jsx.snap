// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect() connects component to store with state 1`] = `
<span>
  Value is: 
  42
</span>
`;

exports[`connect() connects component to store without state 1`] = `
<span>
  No value.
</span>
`;

exports[`connect() does not do anything if no stores are provided 1`] = `
<span>
  Hello World!
</span>
`;

exports[`connect() executes action on store upon mount 1`] = `
<span>
  Value is: 
  23
</span>
`;

exports[`connect() renders spinner when store is not ready 1`] = `
<span>
  <i
    className="fa fa-spin fa-spinner"
  />
   
  Loading...
</span>
`;

exports[`connect() waits on store per default before rendering component when more than zero actions are passed 1`] = `
<span>
  <i
    className="fa fa-spin fa-spinner"
  />
   
  Loading...
</span>
`;
